cache/ringbuffer.go

```diff
@@ -0,0 +1,86 @@
+package cache
+
+import (
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+type entry *clientv3.Event
+
+type ringBuffer struct {
+	mu                sync.RWMutex
+	buffer            []entry
+	head, tail        int
+	oldestRev, latest int64
+}
+
+func newRingBuffer(capacity int) *ringBuffer {
+	// assume capacity > 0 – validated by Cache
+	return &ringBuffer{buffer: make([]entry, capacity)}
+}
+
+func (r *ringBuffer) Append(ev *clientv3.Event) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	rev := ev.Kv.ModRevision
+	r.buffer[r.head] = ev
+	r.head = (r.head + 1) % len(r.buffer)
+
+	if r.head == r.tail { // buffer full -> overwrite oldest
+		r.tail = (r.tail + 1) % len(r.buffer)
+	}
+	r.oldestRev = r.buffer[r.tail].Kv.ModRevision
+	r.latest = rev
+}
+
+func (r *ringBuffer) GetSince(rev int64, pred KeyPredicate) ([]*clientv3.Event, int64) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	if r.tail == r.head || rev > r.latest {
+		return nil, r.latest
+	}
+	out := make([]*clientv3.Event, 0, len(r.buffer))
```

Comment:
Are we over-allocating here?

Reply:
We don't know how much events we will need, and those are cheap pointers, so it's "ok".